crypto/x509/x509_cmp.c

```diff
@@ -196,6 +196,8 @@ int X509_add_cert(STACK_OF(X509) *sk, X509 *cert, int flags)
         ERR_raise(ERR_LIB_X509, ERR_R_PASSED_NULL_PARAMETER);
         return 0;
     }
+    if (cert == NULL)
+        return 1;
```

Comment:
```suggestion
        return 0;
```

Reply:
If you look at my first commit, it was returning 0 and the tests are not passing.
Also in the case of x509_add_certs, certs=NULL is allowed and in this case 1 is returned.

Reply:
The chain (certs) being null, is of course expected a normal. But adding a NULL cert to the chain is quite anomalous. In which case did you find tests failing.

Reply:
If there are tests that expect success adding a NULL cert to a chain, then the tests are wrong!

Reply:
Refering to the docs:

https://docs.openssl.org/master/man3/X509_add_cert/

NULL certs are accepted for `X509_add_certs()` as a no-op from my understanding. `X509_add_certs()` implicitly calls `X509_add_cert()` through `ossl_x509_add_certs_new()`.

I don't have strong opinions on the return code, but I'd like to have consistent behaviour between the two functions. So I think handling NULL cert as a no-op and a success is correct.

Reply:
Let's not confuse the (stack of) `certs` argument, which may indeed, in accordance with the documentation be NULL, with possible single *certs* (values of the single non-stack `cert` argument), which MUST NOT be NULL.

Otherwise, without the up-ref, the code would have pushed NULL objects onto the stack wreaking havoc with later consumers of said stack.

Reply:
I see nothing in the docs to suggest that the `cert` argument is allowed to be NULL>

Reply:
Return success for a NULL `cert` (as opposed to `certs`) would be wrong.

Reply:
I think I misunderstood the docs.

> It does not modify the list certs but in case the X509_ADD_FLAG_UP_REF flag (described below) is set the reference counters of those of its members added to sk are increased.

Seemed like a continuation of this:

> The certs argument may be NULL, which implies no effect.

My bad.

> Return success for a NULL `cert` (as opposed to `certs`) would be wrong.

As long as it is documented it is fine by me!

Reply:
Let's make it zero for a NULL cert, it is an invalid call.